---
title: "Class 10: Structural Bioinformatics (Pt.1)"
author: "Emily Hickey (PID: A15575724)"
format: html
---

## Introduction to the RCSB Protein Data Bank (PDB)

## PDB Statistics 

First, let's see what is in the PDB database - the main repository of protein structures.

Downloaded composition stats from: https://www.rcsb.org/stats/summary

```{r}
stats <- read.csv("Data Export Summary.csv", row.names = 1)
stats
```

There is a problem here due to commas in the numbers. This causes R to treat them as characters.

```{r}
x <- stats$X.ray
x
```

```{r}
as.numeric(gsub(",", "",x))
```

```{r}
rm.comma <- function(x) {
  as.numeric(gsub(",", "",x))
}
```

```{r}
rm.comma(stats$EM)
```

I can use `apply()` to fix the whole table...

```{r}
pdbstats <- apply(stats, 2, rm.comma)
rownames(pdbstats) <- rownames(stats)
head(pdbstats)
```

>Q1. What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
totals <- apply(pdbstats, 2, sum)
round(totals/totals["Total"]*100, 2)
```

>Q2: What proportion of structures in the PDB are protein?

```{r}
round(pdbstats[,"Total"]/sum(pdbstats[,"Total"])*100,2)
```

>Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

We see just one atom per water molecule in this structure because of the resolution. Hydrogen is too small.

>Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have?

308

>Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

Here is a lovely figure of HIP-Pr with the catalytic ASP residues, the MK1 compound and the all important water 308

![](1HSG.png)

## The bio3d package for structural bioinformatics 

```{r}
library(bio3d)
pdb <- read.pdb("1hsg")
```

```{r}
pdb
```

```{r}
head(pdb$atom)
```

## Predicting functional motions of a single structure

Let's finish today with a bioinformatics calculation 
```{r}
adk <- read.pdb("6s36")
m <- nma(adk)
```

```{r}
plot(m)
mktrj(m, file="adk_m7.pdb")
```

>Q7: How many amino acid residues are there in this pdb object?

198

>Q8: Name one of the two non-protein residues?

MK1

>Q9: How many protein chains are in this structure?

2

## Comparative structure analysis of Adenylate Kinase 

We need some packages for todays class. These include `bio3d` and `msa`. 

The `msa` package is from BioConductor. These packages focus on genomics type work and are managed by the `BiocManager` package.

Install `install.packages("BiocManager)` and then `BiocManager::install("msa")` all entered in the R "brain" console. 

```{r}
library(bio3d)

aa <- get.seq("1ake_A")
aa
```

Now I can search the PDB for related sequences: 
```{r}
b <- blast.pdb(aa)
```

```{r}
attributes(b)
hit(b$hits)
```

```{r}
hits <- plot(b)
```

```{r}
hits$pdb.id
```

